### ISSUE TITLE
Accessing paquet file over DataLake Gen 2 using Service Principal is failing

### ISSUE THREAD

**Issue by arnabbiswas1** [dask/adlfs#139]
I am trying to access parquet files from Azure DataLake Gen 2 using service principal (TENANT_ID, CLIENT_ID, CLIENT_SECRET) using the following code:

```
storage_options={'account_name': ACCOUNT_NAME, 'tenant_id': TENANT_ID, "client_id": CLIENT_ID, "client_secret": CLIENT_SECRET}

ddf = dd.read_parquet('az://<container_name>/<path_to_parquet_file>/20201101.parquet', storage_options=storage_options)
```
It fails with the following error:

>TypeError: object AccessToken can't be used in 'await' expression

The [documentation ](https://github.com/dask/adlfs/blob/master/README.md) says:

>Operations against the Gen2 Datalake are implemented by leveraging multi-protocol access, using the Azure Blob Storage Python SDK. The AzureBlobFileSystem accepts all of the BlockBlobService arguments.

As per my knowledge, `BlockBlobService` does not support Service Principal. `BlockBlobService` is part of [v2.1](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python-legacy). v2.1 has currently been declared as legacy and suggested to be replaced by [v12](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python).

The above code works if account_name and account_key is used instead of service principal. **But, I guess, that defeats the purpose of using service principal.**

**Environment**:

- Dask version: 2.30
- Python version: 3.8.3
- Operating System: Ubuntu
- Install method (conda, pip, source): pip

Detailed Stack Trace:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-880d0b4a196b> in <module>
----> 1 ddf = dd.read_parquet(f"az://{CONTAINER_NAME}/{FILE_NAME}", storage_options=storage_options)

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/dask/dataframe/io/parquet/core.py in read_parquet(path, columns, filters, categories, index, storage_options, engine, gather_statistics, split_row_groups, chunksize, **kwargs)
    229         index = [index]
    230 
--> 231     meta, statistics, parts, index = engine.read_metadata(
    232         fs,
    233         paths,

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/dask/dataframe/io/parquet/arrow.py in read_metadata(cls, fs, paths, categories, index, gather_statistics, filters, split_row_groups, **kwargs)
    653             split_row_groups,
    654             gather_statistics,
--> 655         ) = _gather_metadata(
    656             paths, fs, split_row_groups, gather_statistics, filters, dataset_kwargs
    657         )

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/dask/dataframe/io/parquet/arrow.py in _gather_metadata(paths, fs, split_row_groups, gather_statistics, filters, dataset_kwargs)
    164 
    165     # Step 1: Create a ParquetDataset object
--> 166     dataset, base, fns = _get_dataset_object(paths, fs, filters, dataset_kwargs)
    167     if fns == [None]:
    168         # This is a single file. No danger in gathering statistics

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/dask/dataframe/io/parquet/arrow.py in _get_dataset_object(paths, fs, filters, dataset_kwargs)
    134         if proxy_metadata:
    135             dataset.metadata = proxy_metadata
--> 136     elif fs.isdir(paths[0]):
    137         # This is a directory.  We can let pyarrow do its thing.
    138         # Note: In the future, it may be best to avoid listing the

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/adlfs/spec.py in isdir(self, path)
   1050 
   1051     def isdir(self, path):
-> 1052         return maybe_sync(self._isdir, self, path)
   1053 
   1054     async def _isdir(self, path):

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/fsspec/asyn.py in maybe_sync(func, self, *args, **kwargs)
     95         if inspect.iscoroutinefunction(func):
     96             # run the awaitable on the loop
---> 97             return sync(loop, func, *args, **kwargs)
     98         else:
     99             # just call the blocking function

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/fsspec/asyn.py in sync(loop, func, callback_timeout, *args, **kwargs)
     66     if error[0]:
     67         typ, exc, tb = error[0]
---> 68         raise exc.with_traceback(tb)
     69     else:
     70         return result[0]

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/fsspec/asyn.py in f()
     50             if callback_timeout is not None:
     51                 future = asyncio.wait_for(future, callback_timeout)
---> 52             result[0] = await future
     53         except Exception:
     54             error[0] = sys.exc_info()

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/adlfs/spec.py in _isdir(self, path)
   1062                     return True
   1063         try:
-> 1064             info = await self._info(path)
   1065             return info["type"] == "directory"
   1066         except IOError:

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/adlfs/spec.py in _info(self, path, **kwargs)
    507         """
    508         path = self._strip_protocol(path)
--> 509         out = await self._ls(self._parent(path), **kwargs)
    510         out = [o for o in out if o["name"].rstrip("/") == path]
    511         if out:

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/adlfs/spec.py in _ls(self, path, invalidate_cache, delimiter, return_glob, **kwargs)
    685                     outblobs = []
    686                     try:
--> 687                         async for next_blob in blobs:
    688                             if depth in [0, 1] and path == "":
    689                                 outblobs.append(next_blob)

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/async_paging.py in __anext__(self)
    152         if self._page_iterator is None:
    153             self._page_iterator = self.by_page()
--> 154             return await self.__anext__()
    155         if self._page is None:
    156             # Let it raise StopAsyncIteration

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/async_paging.py in __anext__(self)
    155         if self._page is None:
    156             # Let it raise StopAsyncIteration
--> 157             self._page = await self._page_iterator.__anext__()
    158             return await self.__anext__()
    159         try:

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/async_paging.py in __anext__(self)
     97             raise StopAsyncIteration("End of paging")
     98         try:
---> 99             self._response = await self._get_next(self.continuation_token)
    100         except AzureError as error:
    101             if not error.continuation_token:

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/storage/blob/aio/_list_blobs_helper.py in _get_next_cb(self, continuation_token)
     68     async def _get_next_cb(self, continuation_token):
     69         try:
---> 70             return await self._command(
     71                 prefix=self.prefix,
     72                 marker=continuation_token or None,

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/storage/blob/_generated/aio/operations_async/_container_operations_async.py in list_blob_hierarchy_segment(self, delimiter, prefix, marker, maxresults, include, timeout, request_id, cls, **kwargs)
   1322         # Construct and send request
   1323         request = self._client.get(url, query_parameters, header_parameters)
-> 1324         pipeline_response = await self._client._pipeline.run(request, stream=False, **kwargs)
   1325         response = pipeline_response.http_response
   1326 

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in run(self, request, **kwargs)
    213             else _AsyncTransportRunner(self._transport)
    214         )
--> 215         return await first_node.send(pipeline_request)

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in send(self, request)
     81         await _await_result(self._policy.on_request, request)
     82         try:
---> 83             response = await self.next.send(request)  # type: ignore
     84         except Exception:  # pylint: disable=broad-except
     85             if not await _await_result(self._policy.on_exception, request):

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in send(self, request)
     81         await _await_result(self._policy.on_request, request)
     82         try:
---> 83             response = await self.next.send(request)  # type: ignore
     84         except Exception:  # pylint: disable=broad-except
     85             if not await _await_result(self._policy.on_exception, request):

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in send(self, request)
     81         await _await_result(self._policy.on_request, request)
     82         try:
---> 83             response = await self.next.send(request)  # type: ignore
     84         except Exception:  # pylint: disable=broad-except
     85             if not await _await_result(self._policy.on_exception, request):

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in send(self, request)
     81         await _await_result(self._policy.on_request, request)
     82         try:
---> 83             response = await self.next.send(request)  # type: ignore
     84         except Exception:  # pylint: disable=broad-except
     85             if not await _await_result(self._policy.on_exception, request):

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in send(self, request)
     81         await _await_result(self._policy.on_request, request)
     82         try:
---> 83             response = await self.next.send(request)  # type: ignore
     84         except Exception:  # pylint: disable=broad-except
     85             if not await _await_result(self._policy.on_exception, request):

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in send(self, request)
     81         await _await_result(self._policy.on_request, request)
     82         try:
---> 83             response = await self.next.send(request)  # type: ignore
     84         except Exception:  # pylint: disable=broad-except
     85             if not await _await_result(self._policy.on_exception, request):

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_base_async.py in send(self, request)
     79         :rtype: ~azure.core.pipeline.PipelineResponse
     80         """
---> 81         await _await_result(self._policy.on_request, request)
     82         try:
     83             response = await self.next.send(request)  # type: ignore

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/_tools_async.py in await_result(func, *args, **kwargs)
     30     if hasattr(result, "__await__"):
     31         # type ignore on await: https://github.com/python/mypy/issues/7587
---> 32         return await result  # type: ignore
     33     return result

/data/anaconda/envs/azure_dask/lib/python3.8/site-packages/azure/core/pipeline/policies/_authentication_async.py in on_request(self, request)
     35         async with self._lock:
     36             if self._need_new_token:
---> 37                 self._token = await self._credential.get_token(*self._scopes)  # type: ignore
     38         self._update_headers(request.http_request.headers, self._token.token)

TypeError: object AccessToken can't be used in 'await' expression
```

**Comment by hayesgb** [comment]
Which version of alerts are you curre using, and does your code run with adlfs==0.5.5?

**Comment by lossyrob** [comment]
This looks to he the same issue that was fixed in #132

**Comment by arnabbiswas1** [comment]
Awesome! Sorry that I missed that.

Tried with dask/adlfs version 0.5.7 and the above code works. 

May be the documentation should be updated now?

**Comment by hayesgb** [comment]
Closed with #149
